test-e2e/utils.js

```diff
@@ -215,6 +220,179 @@ export function createManager(seed, t, deviceType) {
+ *
+ * The name is deliberately a little obtuse because this class is a little
+ * weird, and we want people to read the documentation.
+ */
+export class ManagerCustodian {
+  #t
+  #rootKey = getRootKey()
+  #dbFolder = temporaryDirectory()
+  #coreStorage = temporaryDirectory()
+
+  /**
+   * @param {import('brittle').TestInstance} t
+   */
+  constructor(t) {
+    this.#t = t
+    for (const folder of [this.#dbFolder, this.#coreStorage]) {
+      t.teardown(() =>
+        fsPromises.rm(folder, { recursive: true, force: true, maxRetries: 2 })
+      )
+    }
+  }
+
+  /**
+   * Run a function on a new manager in a separate process.
+   *
+   * Because this function is serialized and run in a separate process, there
+   * are restrictions on the function:
+   *
+   * 1. Anything referenced outside the function must be passed as an argument.
+   * 2. Arguments and return values must be serializable using `v8.serialize`.
+   *
+   * @template ArgsT
+   * @template ResultT
+   * @param {(manager: MapeoManager, ...args: ArgsT[]) => Promise<ResultT>} fn
+   * @param {ArgsT[]} args
+   * @returns {Promise<ResultT>}
+   */
+  async withManagerInSeparateProcess(fn, ...args) {
+    const jsFile = await this.#createJsFile(fn, args)
+    return this.#runJsFile(jsFile)
+  }
+
+  /**
+   * @template ArgsT
+   * @template ResultT
+   * @param {(manager: MapeoManager, ...args: ArgsT[]) => Promise<ResultT>} fn
+   * @param {ArgsT[]} args
+   * @returns {Promise<string>}
+   */
+  async #createJsFile(fn, args) {
+    // JSON.stringify does a good job wrapping strings in quotes and escaping.
+    // This is a short alias.
+    const s = JSON.stringify
+
+    const require = createRequire(import.meta.url)
+    const fastifyPath = require.resolve('fastify')
+
+    const __filename = fileURLToPath(import.meta.url)
+    const mapeoManagerJsPath = path.resolve(
+      __filename,
+      '..',
+      '..',
+      'src',
+      'mapeo-manager.js'
+    )
+    const rootKeyHex = this.#rootKey.toString('hex')
+
+    const argsHex = v8.serialize(args).toString('hex')
+
+    const source = `
+    import * as v8 from 'node:v8'
+    import Fastify from ${s(fastifyPath)}
+    import { MapeoManager } from ${s(mapeoManagerJsPath)}
+
+    const manager = new MapeoManager({
+      rootKey: Buffer.from(${s(rootKeyHex)}, 'hex'),
+      projectMigrationsFolder: ${s(projectMigrationsFolder)},
+      clientMigrationsFolder: ${s(clientMigrationsFolder)},
+      dbFolder: ${s(this.#dbFolder)},
+      coreStorage: ${s(this.#coreStorage)},
+      fastify: Fastify(),
+    })
+
+    const fn = ${fn.toString()}
+
+    const argsBuffer = Buffer.from(${s(argsHex)}, 'hex')
+    const args = v8.deserialize(argsBuffer)
+
+    const result = await fn(manager, ...args)
+
+    process.send(result)
+    `
+
+    const result = temporaryFile({ extension: 'mjs' })
+    this.#t.teardown(() => fsPromises.rm(result, { maxRetries: 2 }))
```

Comment:
only nit, why is there two calls to `t.teardown`?

Reply:
`dbFolder` and `coreStorage` are created once per `ManagerCustodian`. These JS files are created once per call to `withManagerInSeparateProcess`.

In my test, for example, `dbFolder` is created once (and therefore needs to be removed once) and the JS file is created twice (and therefore needs to be removed twice).

Does that answer your question?